planners/src/solver.rs

```diff
@@ -0,0 +1,269 @@
+    /// When repeated, several strategies will be run in parallel.
+    #[structopt(long = "strategy", short = "s")]
+    strategies: Vec<Strat>,
+}
+
+pub struct Planner {
+    pub option: Opt,
+    pub problem: Option<FiniteProblem>,
+    pub plan: Option<Arc<Domains>>,
+    pub start: Instant,
+    pub end: Instant,
+    pub htn_mode: bool,
+}
+
+impl Planner {
+    pub fn new(option: Opt) -> Self {
+        Self {
+            option,
+            problem: None,
+            plan: None,
+            start: Instant::now(),
+            end: Instant::now(),
+            htn_mode: false,
+        }
+    }
+
+    pub fn get_answer(&self) -> Option<Arc<Domains>> {
+        self.plan.clone()
+    }
+
+    pub fn solve(&mut self, mut spec: Problem, opt: &Opt) -> Result<()> {
+        let mut result = None;
+
+        println!("===== Preprocessing ======");
+        aries_planning::chronicles::preprocessing::preprocess(&mut spec);
+        println!("==========================");
+
+        // if not explicitly given, compute the min/max search depth
+        let max_depth = opt.max_depth.unwrap_or(u32::MAX);
+        let min_depth = if let Some(min_depth) = opt.min_depth {
+            min_depth
+        } else if self.htn_mode && hierarchical_is_non_recursive(&spec) {
+            max_depth // non recursive htn: bounded size, go directly to max
+        } else {
+            0
+        };
+
+        self.start = Instant::now();
+        for n in min_depth..=max_depth {
+            let mut pb = FiniteProblem {
+                model: spec.context.model.clone(),
+                origin: spec.context.origin(),
+                horizon: spec.context.horizon(),
+                chronicles: spec.chronicles.clone(),
+                tables: spec.context.tables.clone(),
+            };
+            let depth_string = if n == u32::MAX {
+                "∞".to_string()
+            } else {
+                n.to_string()
+            };
+            println!("{} Solving with {} actions", depth_string, depth_string);
+            if self.htn_mode {
+                populate_with_task_network(&mut pb, &spec, n)?;
+            } else {
+                populate_with_template_instances(&mut pb, &spec, |_| Some(n))?;
+            }
+            println!("  [{:.3}s] Populated", self.start.elapsed().as_secs_f32());
+            if opt.no_search {
+                propagate_and_print(&pb);
+                break;
+            } else {
+                result = _solve(&pb, opt, self.htn_mode);
+                println!("  [{:.3}s] Solved", self.start.elapsed().as_secs_f32());
+            }
+            if result.is_some() {
+                self.problem = Some(pb.clone());
+                break;
+            }
+        }
+
+        self.end = Instant::now();
+        self.plan = result;
+        Ok(())
+    }
+
+    pub fn format_plan(&self, plan: &Option<Arc<Domains>>) -> Result<()> {
```

Comment:
This performs many side effects and should probably remain in `lcp.rs`. The logic is actually quite simple and different `main` implementation are likely to want different behaviors anyway.

In general, the method name should have been `print_plan` as it does more than just formatting.